website/api/bonsai/bonsai-on-eth.md

```diff
@@ -2,143 +2,97 @@
+An example for sending a callback request via the REST API can be found in the [relay directory of the Bonsai Foundry Template].
 
-    // Send the callback request to the Bonsai Relay.
-    relay_client
-        .callback_request(request)
-        .await
-        .context("Callback request failed")?;
+[REST]: https://en.wikipedia.org/wiki/REST
+[Bonsai Relay SDK]: https://docs.rs/crate/bonsai-ethereum-relay/latest
+[relay directory of the Bonsai Foundry Template]: https://github.com/risc0/bonsai-foundry-template/blob/main/relay/examples/offchain_request.rs
 
-    Ok(())
-}
+### On-chain Requests
 
-```
+The Bonsai Relay also accepts requests on-chain using the [`BonsaiRelay` contract].
 
-### On-chain
+On-chain requests can be useful for reducing the amount of code you need to maintain.
+When submitting requests off-chain, some client or indexer needs to be written to assemble the inputs and make the request to the Bonsai Relay.
+With on-chain requests, the application smart contract can directly issue requests, and so no client or indexer needs to be written or maintained.
 
-As an alternative to sending a `Callback request` from the REST API as described by step 5 of the previous section, the request can be sent via the _on-chain_ `Callback request` interface provided by the [Bonsai Relay Contract]:
+An example of sending a request on-chain can be found in the [starter contract of the Bonsai Foundry Template].
 
-- Send a transaction to Smart Contract `A` to trigger a `Callback request` event that the Bonsai Relay will catch and forward to Bonsai.
+[starter contract of the Bonsai Foundry Template]: https://github.com/risc0/bonsai-foundry-template/blob/main/contracts/BonsaiStarter.sol#L60-L68
 
-#### Example
+## Verifier Contract
 
-Using `cast`:
+RISC Zero supports an on-chain verifier for [Groth16 SNARK] proofs generated by the Bonsai proving service.
+You can use this contract in your applications.
 
-```bash
-cast send --private-key 0x59c6995e998f97a5a0044966f0945389dc9e86dae88c7a8412f4603b6b78690d --gas-limit 100000 "$APP_ADDRESS" 'calculateFibonacci(uint256)' 5
-```
-
-where:
-
-- `$APP_ADDRESS` is an env variable containing the address of your Smart Contract `A`
-- `calculateFibonacci(uint256)` is the function selector of your Smart Contract `A` triggering the `Callback request`
-- `5` is the input for your FIBONACCI program
-
-## Getting Started
-
-The [Bonsai Foundry Template] is the best place to get started building Bonsai applications for Ethereum.
-You may also want to check out our [Bonsai Quick Start](quickstart.md) page.
+As [described above], the Bonsai Relay forwards requests to the Bonsai proving service, and sends proofs on-chain to be verified in the authentication for for the [`BonsaiRelay` contract].
+An alternative to using the Bonsai Relay is to send requests to the Bonsai proving service directly, and to use the verifier contract in your own application contract.
```

Comment:
```suggestion
An alternative to using the Bonsai Relay involves directly sending requests to the Bonsai proving service and incorporating the verifier contract into your own application contract.
```

Reply:
"incorporating the verifier contract into" does not seem quite right to me because I read that phrasing as a requirement to statically link the verifier into the application contract. Static linking is not a requirement since the application can simply call the verifier contract with an EVM message call.